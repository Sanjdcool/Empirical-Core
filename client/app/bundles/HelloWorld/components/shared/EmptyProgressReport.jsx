import React from 'react';

export default React.createClass({

  render() {
    let link,
      buttonText,
      content;
    if (this.props.missing === 'activities') {
      link = '/teachers/classrooms/activity_planner/assign-new-activity';
      buttonText = 'Assign an Activity';
      content = 'In order to access our different reports, you need to assign activities to your students.';
    } else if (this.props.missing === 'students') {
      link = '/teachers/classrooms/invite_students';
      buttonText = 'Invite Students';
      content = 'In order to access our different reports, you need to invite your students and assign activities.'
    } else {
      link = '/teachers/classrooms/new';
      buttonText = 'Create a Class';
      content = 'In order to access our different reports, you need to create a class and assign activities to your students.'
    }
 		return (
   <div className="empty-progress-report">
     <img src='/images/empty_state_illustration.png' />
     <h1>You have no reports yet!</h1>
     <p>{content}</p>
     <button onClick={() => { window.location = link; }} className="button-green create-unit featured-button">{buttonText}</button>
     <a href="/teacher_resources">Teacher Resources</a>
     {/* <div className="row empty-state-manager">
       <div className="col-xs-7">
         {content}
       </div>
       <div className="col-xs-4">
         <button onClick={() => { window.location = link; }} className="button-green create-unit featured-button">{buttonText}</button>
       </div>
     </div> */}
   </div>
 );
  },
});
