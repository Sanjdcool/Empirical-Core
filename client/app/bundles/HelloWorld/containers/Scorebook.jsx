import React from 'react';
import Scrollify from '../components/modules/scrollify';
import $ from 'jquery';
import _ from 'underscore';
import TableFilterMixin from '../components/general_components/table/sortable_table/table_filter_mixin.js';
import StudentScores from '../components/scorebook/student_scores';
import LoadingIndicator from '../components/shared/loading_indicator';
import ScorebookFilters from '../components/scorebook/scorebook_filters';
import ScoreLegend from '../components/scorebook/score_legend';
import AppLegend from '../components/scorebook/app_legend.jsx';
import EmptyProgressReport from '../components/shared/EmptyProgressReport';

export default React.createClass({
  mixins: [TableFilterMixin],

  getInitialState() {
    this.modules = {
      scrollify: new Scrollify(),
    };
    return {
      units: [],
      classrooms: this.props.allClassrooms,
      selectedUnit: {
        name: 'All Activity Packs',
        value: '',
      },
      selectedClassroom: this.selectedClassroom(),
      classroomFilters: [],
      unitFilters: [],
      scores: {},
      beginDate: null,
      premium_state: this.props.premium_state,
      endDate: null,
      currentPage: 0,
      loading: false,
      is_last_page: false,
      noLoadHasEverOccurredYet: true,
    };
  },

  componentDidMount() {
    this.fetchData();
    this.modules.scrollify.scrollify('#page-content-wrapper', this);
  },

  formatDate(date) {
    if (date) {
      return `${date.year()}-${date.month() + 1}-${date.date()}`;
    }
  },

  fetchData() {
    const newCurrentPage = this.state.currentPage + 1;
    this.setState({ loading: true, currentPage: newCurrentPage, });
    $.ajax({
      url: '/teachers/classrooms/scores',
      data: {
        current_page: newCurrentPage,
        classroom_id: this.state.selectedClassroom.value,
        unit_id: this.state.selectedUnit.value,
        begin_date: this.formatDate(this.state.beginDate),
        end_date: this.formatDate(this.state.endDate),
        selectedClassroom: this.state.selectedClassroom.id,
        no_load_has_ever_occurred_yet: this.state.noLoadHasEverOccurredYet,
      },
      success: this.displayData,
    });
  },

  displayData(data) {
    this.setState({
      classroomFilters: this.getFilterOptions(this.state.classrooms, 'name', 'id', 'Select a Classroom'),
      unitFilters: this.getFilterOptions(data.units, 'name', 'id', 'All Activity Packs'),
      is_last_page: data.is_last_page,
      premium_state: data.premium_state,
      noLoadHasEverOccurredYet: false,
    });
    const newScores = Object.assign({}, this.state.scores);
    data.scores.forEach((s) => {
      // add the score to the user scores arr or create a new one
      newScores[s.user_id] = newScores[s.user_id] || { name: s.name, scores: [], };
      newScores[s.user_id].scores.push({
        caId: s.ca_id,
        // activitySessionId: s.id,
        userId: s.user_id,
        updated: s.updated_at,
        name: s.activity_name,
        percentage: s.percentage,
        activity_classification_id: s.activity_classification_id, });
    });
    this.setState({ loading: false, scores: newScores, });
  },

  selectedClassroom() {
    if (!this.props.selectedClassroom) {
      return { name: 'None Selected', value: 'none_selected', };
    }
    return this.props.selectedClassroom;
  },

  selectUnit(option) {
    this.setState({
      scores: {},
      currentPage: 0,
      selectedUnit: option,
    }, this.fetchData);
  },

  selectClassroom(option) {
    this.setState({
      scores: {},
      currentPage: 0,
      selectedClassroom: option,
    }, this.fetchData
  );
  },

  selectDates(val1, val2) {
    this.setState({
      scores: {},
      currentPage: 0,
      beginDate: val1,
      endDate: val2,
    }, this.fetchData);
  },

  render() {
    let content,
      loadingIndicator;
    const scores = [];
    for (const userId in this.state.scores) {
      if (this.state.scores.hasOwnProperty(userId)) {
        const sObj = this.state.scores[userId];
        scores.push(<StudentScores key={userId} data={{ scores: sObj.scores, name: sObj.name, activity_name: sObj.activity_name, userId: sObj.userId, }} premium_state={this.props.premium_state} />);
      }
    }
    if (this.state.loading) {
      loadingIndicator = <LoadingIndicator />;
    } else {
      loadingIndicator = null;
    }

    if (this.props.missing) {
      content = <EmptyProgressReport missing={this.props.missing} />;
    } else {
      content = (<span>
        <div className="container">
          <section className="section-content-wrapper">
            <ScorebookFilters selectedClassroom={this.state.selectedClassroom} classroomFilters={this.state.classroomFilters} selectClassroom={this.selectClassroom} selectedUnit={this.state.selectedUnit} unitFilters={this.state.unitFilters} selectUnit={this.selectUnit} selectDates={this.selectDates} />
            <ScoreLegend />
            <AppLegend />
          </section>
        </div>
        {scores}
        {loadingIndicator}
      </span>);
    }
    return (
      <div className="page-content-wrapper">
        <div className="tab-pane" id="scorebook">
          {content}
        </div>
      </div>
    );
  },
});
