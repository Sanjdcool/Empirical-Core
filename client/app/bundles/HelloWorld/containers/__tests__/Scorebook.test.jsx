import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import Scorebook from '../Scorebook.jsx'
import EmptyProgressReport from '../../components/scorebook/EmptyProgressReport'
import LoadingIndicator from '../../components/general_components/loading_indicator'
import StudentScores from '../../components/scorebook/student_scores'
import ScorebookFilters from '../../components/scorebook/scorebook_filters'
import ScoreLegend from '../../components/scorebook/score_legend'
import AppLegend from '../../components/scorebook/app_legend.jsx'

const classrooms = [{ name: 'A', id: 1}, { name: 'B', id: 2}, { name: 'C', id: 3}]
const units = [{ name: 'Something', id: 4}, { name: 'Someone', id: 5}, { name: 'Somewhere', id: 6}]
const scores = []

for (let x=0; x<30; x++) {
  scores.push({results: '', user: {id: x}})
}

const data = {
  teacher: { premium_state: 'trial' },
  is_last_page: false,
  classrooms,
  units,
  scores
}

describe('Scorebook component', () => {

  it('should render', () => {
    const wrapper = shallow(<Scorebook />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('if props.missing has a value', () => {
    const wrapper = shallow(<Scorebook missing={'students'}/>)

    it('renders EmptyProgressReport', () => {
      expect(wrapper.find(EmptyProgressReport).length).toEqual(1)
    })

    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0)
    })

    it('does not render ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(0)
    })

    it('does not render ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(0)
    })

    it('does not render AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(0)
    })

    it('does not render StudentScores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(0)
    })


    // it('does not get rendered if it is not passed a value for missing', () => {
    //   const wrapper = shallow(<Scorebook missing={''}/>)
    //   expect(wrapper.find(EmptyProgressReport).length).toEqual(0)
    // })
  })

  describe('if state.loading is true', () => {
    const wrapper = shallow(<Scorebook/>)
    wrapper.setState({loading: true})

    it('renders LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(1)
    })

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1)
    })

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1)
    })

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1)
    })

    it('does not render StudentScores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(0)
    })
  })

  describe('if it gets data when it had none', () => {

    const wrapper = shallow(<Scorebook/>)
    wrapper.setState({currentPage: 1})
    wrapper.instance().displayData(data)

    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0)
    })

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1)
    })

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1)
    })

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1)
    })

    it('updates the state.scores to be equal to the new ones', () => {
      expect(wrapper.state('scores')).toEqual(scores)
    })

    it('renders as many StudentScores as there are state.scores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(wrapper.state('scores').length)
    })
  })

  describe('if it gets data when it had some', () => {

    const wrapper = shallow(<Scorebook/>)
    const scoreArray = [{results: '', user: {id: 666}}]
    wrapper.setState({scores: scoreArray})
    wrapper.instance().displayData(data)

    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0)
    })

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1)
    })

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1)
    })

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1)
    })

    it('updates the state.scores to add the new ones', () => {
      expect(wrapper.state('scores')).toEqual(scoreArray.concat(scores))
    })

    it('renders as many StudentScores as there are state.scores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(wrapper.state('scores').length)
    })
  })

  describe('selectedClassroom function', () => {

    it('returns All Classrooms if there is no selectedClassroom in props', () => {
      const wrapper = shallow(<Scorebook/>)
      expect(wrapper.instance().selectedClassroom()).toEqual({name: 'All Classrooms', value: ''})
    })

    it('returns props.selectedClassrom if there is a selectedClassroom in props', () => {
      const wrapper = shallow(<Scorebook selectedClassroom={classrooms[0]}/>)
      expect(wrapper.instance().selectedClassroom()).toEqual(classrooms[0])
    })
  })

  it('sets state.selectedUnit when selectUnit is called', () => {
    const wrapper = shallow(<Scorebook/>)
    wrapper.instance().selectUnit(units[0])
    expect(wrapper.state('selectedUnit')).toEqual(units[0])
  })

  it('sets state.selectedClassroom when selectClassroom is called', () => {
    const wrapper = shallow(<Scorebook/>)
    wrapper.instance().selectClassroom(classrooms[0])
    expect(wrapper.state('selectedClassroom')).toEqual(classrooms[0])
  })

  it('sets state.beginDate and state.endDate when selectDates is called', () => {
    const beginDate = moment(-7)
    const endDate = moment(7)
    const wrapper = shallow(<Scorebook/>)
    wrapper.instance().selectDates(beginDate, endDate)
    expect(wrapper.state('beginDate')).toEqual(beginDate)
    expect(wrapper.state('endDate')).toEqual(endDate)
  })


})
