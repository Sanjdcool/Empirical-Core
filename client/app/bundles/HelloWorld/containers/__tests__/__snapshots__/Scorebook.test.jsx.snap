// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scorebook component should render 1`] = `
<div
  className="page-content-wrapper"
>
  <div
    className="tab-pane"
    id="scorebook"
  >
    <span>
      <div
        className="container"
      >
        <section
          className="section-content-wrapper"
        >
          <scorebook_filters
            classroomFilters={Array []}
            selectClassroom={[Function]}
            selectDates={[Function]}
            selectUnit={[Function]}
            selectedClassroom={
              Object {
                "name": "All Classrooms",
                "value": "",
              }
            }
            selectedUnit={
              Object {
                "name": "All Activity Packs",
                "value": "",
              }
            }
            unitFilters={Array []}
          />
          <score_legend />
          <app_legend />
        </section>
      </div>
    </span>
  </div>
</div>
`;
